### PR TITLE
Translate: Select translate server on UI

### PR DESCRIPTION
**Pull request by muxueqz**
This is based on https://github.com/muxueqz/koreader/tree/feat-youdao-translator
Should I merge these to one Pull Request?

---
This change is [Reviewable](https://reviewable.io/reviews/koreader/koreader/9045)

### CODE REVIEW

**Comment by poire-z** [comment]
A few questions, just to be sure it's really needed to have a user choice of the translation server:
- what is "youdao", is it letigitmate, known to be good (for everyotherlanguage > chinese I guess?)
- is google really worse than youdao ?
- does youdao works as well with other translations like French <> Spanish ? Does it support them all, and autodetects source language?

My idea is that if it is really legitimate and known to be better than google when Chinese is the source or target language, just use it instead of Google - don't let users have to decide.
(And if Google is good enough, may be don't add it - less things to maintain and risk of broken features in the future that we won't be able to maintain.)

**Comment by poire-z** [comment]
Closing because lack of answers/activity.

**Comment by junrenshi** [comment]
Actually, the modification is useful not because Youdao is superior than Goggle Translate, but because Translate functionality cannot be used inside mainland China where Great Firewall (GFW) blocks all Goggle domain names. The same GFW also prevents using the Wikipedia functionality.

**Comment by Frenzie** [comment]
I do agree with @poire-z. An extra maintenance burden is generally undesirable. But given this new information combined with the fact that the code looks clean and self-contained, I'll reopen the PR to see if it still applies to master.

**Comment by Frenzie** [comment]
For me neither of the URLs work by the way, but maybe it doesn't outside of China or only specific language combinations are supported?

```
01/26/24-19:50:41 DEBUG Calling https://translate.google.cn/translate_a/single?otf=1&sl=auto&client=gtx&oe=UTF-8&tl=nl_nl&hl=en&ssel=0&tsel=0&dt=t&dt=at&dt=md&ie=UTF-8&q=playing%20company 
01/26/24-19:50:41 WARN  translator HTTP status not okay: HTTP/1.1 404 Not Found 
01/26/24-19:50:41 DEBUG Response headers: {
  ["alt-svc"] = "h3=\":443\"; ma=2592000,h3-29=\":443\"; ma=2592000",
  connection = "close",
  ["content-length"] = "1736",
  ["content-type"] = "text/html; charset=UTF-8",
  date = "Fri, 26 Jan 2024 18:50:41 GMT",
  server = "sffe",
  ["x-content-type-options"] = "nosniff",
  ["x-xss-protection"] = "0"
} --[[table: 0x7f3634d4fb28]] 
```

```
01/26/24-19:51:17 WARN  translator error: nil 
error in subprocess:    frontend/ui/translator.lua:482: attempt to index local 'youdao_result' (a nil value)
stack traceback:
        frontend/ui/translator.lua:482: in function 'task'
        frontend/ui/trapper.lua:562: in function 'func'
        ./ffi/util.lua:374: in function <./ffi/util.lua:345>
        [C]: in function 'xpcall'
        ./ffi/util.lua:345: in function 'runInSubProcess'
        frontend/ui/trapper.lua:547: in function 'dismissableRunInSubprocess'
        frontend/ui/translator.lua:676: in function '_showTranslation'
        frontend/ui/translator.lua:663: in function <frontend/ui/translator.lua:662>
        [C]: in function 'xpcall'
        frontend/ui/trapper.lua:48: in function <frontend/ui/trapper.lua:45>
```

File: frontend/ui/translator.lua [hunk 1]
```diff
@@ -273,6 +291,14 @@ This is useful:
                 sub_item_table = genLanguagesItems("translator_to_language", self:getTargetLanguage()),
                 keep_menu_open = true,
             },
+            {
+                text_func = function()
+                    local value = self:getTransServer()
+                    return T(_("Translate Server: %1"), value)
```
Comment (Frenzie) [inline frontend/ui/translator.lua:297]:
```suggestion
                    return T(_("Translation server: %1"), value)
```

**Comment by poire-z** [comment]
Mixed feelings, and I surely don't want this PR in as it is (more on that below).

I also haven't got any answer about my questions, mainly: who is "youdao" ?! Some official Chinese government owned website, some big chinese company (like Google) service ? Some chinese developper pet project ? Hosted in mainland China or elsewhere?
(I'm not certain we really have the rights to use the Google url we use as we do, but we know there are other projects that use it, and our project is confidential enough, and our user base small enough, that Google won't probably care. No idea about how all that is with Youdao, and I honnestly don't want to know and have to care :) But other people in this project more at ease with licence and free software/services stuff may have some thoughts ?)

(Also, why workaround the issue? Just go fix the issue and revolt and take done that GFW ! :))

About the implementation:
- I like that you somehow at the end masquerade the result and make it look exactly like the object 
we get from Google translate.
- But the function that request stuff from youdao is indeed self-contained, but nothing described like the code we have for Google where we set and describe all the parameters, and build the query string from them. For youdao, you have the full url put as is. I don't have any preferences between the 2 ways, but for our code to look clean, it should be the same way for Google and Youdao.
- So, the youdao code is how I would write it if I made a user patch. And it could be made a user patch - so Chinese users could just download it and drop it, and we main maintainers would not have to care and see it that much integrated (but only half-integrated) in our code base.
- (there is also this `local function http_request(...)...` that is put there (and looks like some common thing we probably already have elsewhere) and it is only used by the youdao code - if it needs to be put there, it should be shared and used by the google code. If it would be in a user patch, I would not care.)
- If you would really want to have this "patching/replacement" of the Google translate code in our main codebase, make it really behave just like a user patch: a single checkbox in Translate settings `[ ] Use youdao instead of Google`, a single added function, fully contained, put at the bottom (or may be in another file) that would be plugged to replace ` Translator:loadPage(text, target_lang, source_lang)`. Google translate would stay the first class citizen, and your quick hack would just replace it when you check that checkbox. We wouldn't see among our default params in Translator.trans_servers hostnames like youdao.com where the other params defined around it won't work with.
- I thought we could just get rid of translate_servers (as there's only one server, and http://translate.google.cn does not work anymore, so it could be harcoded) and get rid of having it configurable and fetched from settings - but I think it's still nice to be able to modify it and use for example services like `https://proxymyurl.com/translate.googleapis.com/` (totally fake url!) where the rest would be appended, if you could not access it directly (I'm quite certain there are ways like this to escape the GFW :)) Anyway, this could be made a user-patch, so no real need to have it in G_reader_settings).

**Comment by Frenzie** [comment]
> * So, the youdao code is how I would write it if I made a user patch. And it could be made a user patch - so Chinese users could just download it and drop it, and we main maintainers would not have to care and see it that much integrated (but only half-integrated) in our code base.

That's an excellent point. I think that may not have been an option yet at the time. Also we could put in place something to facilitate that better.

**Comment by Frenzie** [comment]
Closing based on the rationale in https://github.com/koreader/koreader/pull/9045#issuecomment-1912612782

This could work very well as a user patch, but for now we'd prefer not to add it to the main code.

**Pull request closed without merging**